### PR TITLE
OHRM5X-2177: Add Claim Module to Module configuration

### PR DESCRIPTION
**Pull request by DulsaraNethmin**

### CODE REVIEW

File: src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php [hunk 1]
```diff
@@ -49,7 +49,7 @@ protected function setUp(): void
     public function testGetModuleList(): void
     {
         $moduleList = $this->moduleDao->getModuleList();
-        $this->assertEquals(7, count($moduleList));
+        $this->assertEquals(8, count($moduleList));
```
Comment (ManulMax) [inline src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php:52]:
```suggestion
        $this->assertCount(8, $moduleList);
```

File: src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php [hunk 2]
```diff
@@ -69,7 +69,7 @@ public function testUpdateModuleStatusForReturnedValues(): void
     {
         $moduleUpdateArray = ['admin' => true, 'pim' => false];
         $returedObjects = $this->moduleDao->updateModuleStatus($moduleUpdateArray);
```
Comment (ManulMax) [inline src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php:71]: seems like having spelling mistake in previous code. better to fix 
```suggestion
       $returnedObjects = $this->moduleDao->updateModuleStatus($moduleUpdateArray);
```

File: src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php [hunk 3]
```diff
@@ -69,7 +69,7 @@ public function testUpdateModuleStatusForReturnedValues(): void
...
-        $this->assertEquals(7, count($returedObjects));
+        $this->assertEquals(8, count($returedObjects));
```
Comment (ManulMax) [inline src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php:72]: .

File: src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php [hunk 4]
```diff
@@ -78,7 +78,7 @@ public function testUpdateModuleStatusWithModulesDoesNotExists(): void
     {
         $moduleUpdateArray = ['admin' => true, 'pim' => false, 'test' => true];
         $returedObjects = $this->moduleDao->updateModuleStatus($moduleUpdateArray);
-        $this->assertEquals(7, count($returedObjects));
+        $this->assertEquals(8, count($returedObjects));
```
Comment (ManulMax) [inline src/plugins/orangehrmAdminPlugin/test/Dao/ModuleDaoTest.php:81]: .

**Review by ManulMax — CHANGES_REQUESTED** [review]
Address review  feedback

File: src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php [hunk 5]
```diff
@@ -277,6 +281,10 @@ public function getValidationRuleForUpdate(): ParamRuleCollection
                 self::PARAMETER_DIRECTORY,
                 new Rule(Rules::BOOL_TYPE),
             ),
+            new ParamRule(
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php:284]: is it possible for me to send without `claim` parameter. We have to keep backward compatible.
@ManulMax check this. I have added that to the ticket as well
Basically `claim` param should be optional and it should have a default value maybe `false`

File: src/plugins/orangehrmAdminPlugin/test/fixtures/testcases/ModulesAPITestCases.yaml [hunk 6]
```diff
@@ -13,7 +13,8 @@ GetAll:
       performance: true,
       maintenance: false,
       mobile: false,
-      directory: true
+      directory: true,
+      claim: true
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/test/fixtures/testcases/ModulesAPITestCases.yaml:17]:
```suggestion
```
this should pass without sending `claim` param. but response body will contain `claim` param.

File: src/plugins/orangehrmAdminPlugin/test/fixtures/testcases/ModulesAPITestCases.yaml [hunk 7]
```diff
@@ -37,15 +38,16 @@ Update:
       performance: false,
       maintenance: false,
       mobile: false,
-      directory: false
+      directory: false,
+      claim: false
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/test/fixtures/testcases/ModulesAPITestCases.yaml:42]:
```suggestion
```
same here. All existing API tests request body should not be changed. Write new test cases for cover claim.

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php [hunk 8]
```diff
@@ -221,7 +225,11 @@ public function update(): EndpointResourceResult
     {
         $modules = self::CONFIGURABLE_MODULES;
         foreach ($modules as $key => $module) {
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php:227]:
```suggestion
         foreach (self::CONFIGURABLE_MODULES as $key => $module) {
```
:thinking:

File: src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php [hunk 9]
```diff
@@ -221,7 +225,11 @@ public function update(): EndpointResourceResult
...
-            $modules[$key] = $this->getRequestParams()->getBoolean(RequestParams::PARAM_TYPE_BODY, $key, true);
+            if ($key === self::PARAMETER_CLAIM) {
+                $modules[$key] = $this->getRequestParams()->getBoolean(RequestParams::PARAM_TYPE_BODY, $key, false);
+            } else {
+                $modules[$key] = $this->getRequestParams()->getBoolean(RequestParams::PARAM_TYPE_BODY, $key, true);
+            }
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php:232]: Seems we can revert this as previous

```suggestion
            $modules[$key] = $this->getRequestParams()->getBoolean(RequestParams::PARAM_TYPE_BODY, $key, self::CONFIGURABLE_MODULES[$key] );
```
why can't we do like this.

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php (hunk 8 above)
Comment (DulsaraNethmin) [reply src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php:227]: we have to keep $modules variable, cause we cannot update the content of a constant, self::CONFIGURABLE_MODULES

File: src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php [hunk 10]
```diff
@@ -65,6 +66,7 @@ class ModulesAPI extends Endpoint implements CrudEndpoint
         self::PARAMETER_MAINTENANCE => false,
         self::PARAMETER_MOBILE => false,
         self::PARAMETER_DIRECTORY => false,
+        self::PARAMETER_CLAIM => false,
```
Comment (RajithaKumara) [inline src/plugins/orangehrmAdminPlugin/Api/ModulesAPI.php:69]: I feel this and all other modules' default values should go as `true`. but I don't this we can change this now as its BC break

**Review by RajithaKumara — APPROVED** [review]

**Pull request merged**